### PR TITLE
SW merging

### PR DESCRIPTION
**Pull request by aburrell**
Added merging routines to allow people to combine measured and forecasted Kp and F10.7 indexes into a single instrument object.  Measured values are given priority over forecasted values.

Builds off of pull request #138, which should be merged first.

### CODE REVIEW

**Comment by aburrell** [comment]
@rstoneback , I don't think the new test failures should be happening, but I also don't know how to fix them.

File: pysat/instruments/sw_f107.py [hunk 1]
```diff
@@ -270,15 +324,18 @@ def download(date_array, tag, sat_id, data_path, user=None, password=None):
...
-                     for time in data.pop('time')]
-            data.index = times
-            # replace fill with NaNs
-            idx, = np.where(data['f107'] == -99999.0)
-            data.iloc[idx,:] = np.nan
-            # create file
-            data.to_csv(os.path.join(data_path, 'f107_monthly_' +
-                                     date.strftime('%Y-%m') + '.txt'))
+            if data.empty:
+                print("WARNING: no data for {:}".format(date))
```
Comment (rstoneback) [inline pysat/instruments/sw_f107.py]: Should this be an official warning, like raise RuntimeWarning()

File: pysat/instruments/sw_f107.py [hunk 2]
```diff
@@ -357,27 +440,151 @@ def download(date_array, tag, sat_id, data_path, user=None, password=None):
...
+        # Format the data values
+        for i,kk in enumerate(val_keys):
+            val = split_line[i + 3]
+            if kk != 'goes_bgd_flux':
+                val = int(val)
+            values[kk].append(val)
+
+    return dates, values
+
+def calc_f107a(f107_inst, f107_name='f107', f107a_name='f107a', min_pnts=41):
```
Comment (rstoneback) [inline pysat/instruments/sw_f107.py]: 👍

**Comment by rstoneback** [comment]
I hit the wrong button. I am looking into the tests.

File: pysat/instruments/sw_f107.py (hunk 1 above)
Comment (aburrell) [reply pysat/instruments/sw_f107.py]: If you like, I've not used the official Warnings before.

File: pysat/instruments/sw_f107.py (hunk 2 above)
Comment (aburrell) [reply pysat/instruments/sw_f107.py]: Yeah, only problem is that I need to add in the provisional source of F10.7 now, because we have a big data gap!

**Comment by aburrell** [comment]
> I hit the wrong button. I am looking into the tests.
--

Hehe, good!  I was just thinking:  "I just added an incomplete routine, I don't approve of this pull request anymore!"

File: pysat/tests/test_sw.py [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+import datetime as dt
+
+class TestSWKp():
+    def setup(self):
+        """Runs before every method to create a clean testing setup"""
+        # Load a test instrument
+        self.testInst = pysat.Instrument()
+        self.testInst.data = pds.DataFrame({'Kp': np.arange(0, 4, 1.0/3.0)},
+                                             index=[pysat.datetime(2009, 1, 1)
+                                                    + pds.DateOffset(hours=3*i)
```
Comment (rstoneback) [inline pysat/tests/test_sw.py]: I believe this should be + pds.DateOffset(hours=2*i)

File: pysat/tests/test_sw.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+                                                    for i in range(10)])
```
Comment (rstoneback) [inline pysat/tests/test_sw.py]: for i in range(12)])

File: pysat/tests/test_sw.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+    def test_combine_f107_none(self):
+        """ Test combine_f107 failure when no input is provided"""
+        
+        assert_raises(TypeError, sw_methods.combine_f107)
+
+    def test_combine_f107_no_time(self):
+        """Test combine_f107 failure when no times are provided"""
+
+        assert_raises(ValueError, sw_methods.combine_f107,
+                      [self.combineInst[''], self.combineInst['forecast']])
```
Comment (rstoneback) [inline pysat/tests/test_sw.py]: self.combineInst[''], self.combineInst['forecast'])

each argument to combine_f107 needs to be on its own in the assert raises call

**Comment by rstoneback** [comment]
pyglow is also causing test issues, for reasons that will have to be determined.

File: pysat/tests/test_sw.py (hunk 4 above)
Comment (aburrell) [reply pysat/tests/test_sw.py]: Fixed.

File: pysat/tests/test_sw.py (hunk 5 above)
Comment (aburrell) [reply pysat/tests/test_sw.py]: Fixed.

**Comment by aburrell** [comment]
Added the new data source, but need to figure out how to get the file dating done properly.  This is extra complicated since some files are 3 months long and others are a year long.

**Comment by aburrell** [comment]
@jklenzing , the newest source of F10.7 data is ftp.  Should sw_f107 also be added to the exclude list in the test_instruments.py?

File: pysat/instruments/sw_f107.py (hunk 2 above)
Comment (aburrell) [reply pysat/instruments/sw_f107.py]: Added!

File: pysat/instruments/sw_f107.py (hunk 1 above)
Comment (aburrell) [reply pysat/instruments/sw_f107.py]: Changed to UserWarning

**Comment by aburrell** [comment]
Ok, this needs another pair of eyes now @rstoneback

**Comment by jklenzing** [comment]
> @jklenzing , the newest source of F10.7 data is ftp. Should sw_f107 also be added to the exclude list in the test_instruments.py?

That should probably go in the list.  Is it a replacement, or additional source of data?  Not sure if we can parse the tests so that only the ftp sources are skipped.  @rstoneback?

**Comment by rstoneback** [comment]
I added a way to exclude specific module name, tag, and sat_id combinations from the Travis CI download testing. The code is in Jeff's pull.

**Comment by aburrell** [comment]
Ok, I guess there will be a pull request after this one that adds the exception

File: pysat/instruments/sw_f107.py [hunk 6]
```diff
@@ -521,14 +681,28 @@ def parse_daily_solar_data(data_lines):
...
+        dates.append(pysat.datetime.strptime(" ".join(split_line[0:3]), dfmt))
 
         # Format the data values
+        j = 0
         for i,kk in enumerate(val_keys):
-            val = split_line[i + 3]
+            if year == 1994 and kk == 'new_reg':
+                # New regions only in files after 1994
+                val = -999
+            elif((year == 1994 and kk in xray_keys) or
```
Comment (rstoneback) [inline pysat/instruments/sw_f107.py]: missing space?
elif ((

Comment (aburrell) [reply pysat/instruments/sw_f107.py]: I don't think so:

> Avoid extraneous whitespace in the following situations: Immediately inside parentheses, brackets or braces.

Comment (aburrell) [reply pysat/instruments/sw_f107.py]: Oh, you're talking about a space before.  Ew.  I hope that isn't PEP8!

**Comment by rstoneback** [comment]
Hehehe. I hope it is clear by now that I should never be put in charge of PEP8 compliance.

On Feb 7, 2019, at 10:17 AM, Angeline Burrell <notifications@github.com<mailto:notifications@github.com>> wrote:


@aburrell commented on this pull request.

________________________________

In pysat/instruments/sw_f107.py<https://github.com/rstoneback/pysat/pull/139#discussion_r254753980>:

>          for i,kk in enumerate(val_keys):
-            val = split_line[i + 3]
+            if year == 1994 and kk == 'new_reg':
+                # New regions only in files after 1994
+                val = -999
+            elif((year == 1994 and kk in xray_keys) or


Oh, you're talking about a space before. Ew. I hope that isn't PEP8!

**Comment by aburrell** [comment]
@rstoneback , finally looks like all failures are due to downloads of some nature!  You should verify that before we merge this.

File: pysat/instruments/sw_f107.py [hunk 7]
```diff
@@ -127,31 +130,41 @@ def load(fnames, tag=None, sat_id=None):
     if tag == '45day':
         meta['ap'] = {meta.name_label: 'Daily Ap index',
                       meta.desc_label: 'Daily average of 3-h ap indices'}
-    elif tag == 'daily':
+    elif tag == 'daily' or tag == 'prelim':
         meta['ssn'] = {meta.name_label: 'Sunspot Number',
-                       meta.desc_label: 'SESC Sunspot Number'}
+                       meta.desc_label: 'SESC Sunspot Number',
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/instruments/sw_f107.py]: Yes, they are working just fine.

File: pysat/instruments/sw_f107.py [hunk 8]
```diff
@@ -215,15 +228,52 @@ def list_files(tag=None, sat_id=None, data_path=None, format_str=None):
...
+            out = pysat.Files.from_os(data_path=data_path,
+                                      format_str=format_str)
+
+            if not out.empty:
+                # Set each file's valid length at a 1-day resolution
+                orig_files = out.sort_index().copy()
+                new_files = list()
+
+                for orig in orig_files.iteritems():
+                    # Version determines each file's valid length
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/instruments/sw_f107.py]: Yes, there's an overlapping period where they've created the year long file but still have the quarterly files available.

File: pysat/utils.py [hunk 9]
```diff
@@ -475,6 +475,57 @@ def parse_date(str_yr, str_mo, str_day, str_hr='0', str_min='0', str_sec='0',
...
+        Datetime list, array, or Index
+
+    Returns
+    -------
+    freq : (str)
+       Frequency string as described in Pandas Offset Aliases
+
+    Notes
+    -----
+    Calculates the minimum time difference and sets that as the frequency.
```
[the remark this answers is not in the document]
Comment (aburrell) [reply pysat/utils.py]: This would be good to put in as an enhancement issue.

**Comment by rstoneback** [comment]
Tests confirmed.

**Pull request merged**